cyipopt/tests/unit/test_scipy_optional.py

```diff
@@ -131,6 +135,152 @@ def test_minimize_ipopt_jac_hessians_constraints_with_arg_kwargs():
+        assert d == d0
+        fun_constraint.count += 1
+        return [(x[2] - c) ** 2, (x[3] - d) ** 2]
+
+    def grad_constraint(x, c=0, e=0, d=0):
+        assert c == c0
+        assert d == d0
+        grad_constraint.count += 1
+        return np.hstack((np.zeros((2, 2)),
+                         np.diag([2 * (x[2] - c), 2 * (x[3] - d)])))
+
+    def callback(*args, **kwargs):
+        callback.count += 1
+
+    constr = {
+        "type": "eq",
+        "fun": fun_constraint,
+        "jac": grad_constraint,
+        "args": (c0,),
+        "kwargs": {'d': d0},
+    }
+
+    kwargs = {}
+    jac_methods = {'cg', 'bfgs', 'newton-cg', 'l-bfgs-b', 'tnc', 'slsqp,',
+                   'dogleg', 'trust-ncg', 'trust-krylov', 'trust-exact',
+                   'trust-constr'}
+    hess_methods = {'newton-cg', 'dogleg', 'trust-ncg', 'trust-krylov',
+                   'trust-exact', 'trust-constr'}
+    hessp_methods = hess_methods - {'dogleg', 'trust-exact'}
+    constr_methods = {'slsqp', 'trust-constr'}
+
+    if method in jac_methods:
+        kwargs['jac'] = grad
+    if method in hess_methods:
+        kwargs['hess'] = hess
+    if method in constr_methods:
+        kwargs['constraints'] = constr
+    if method in MINIMIZE_METHODS:
+        kwargs['callback'] = callback
+
+    fun.count = 0
+    grad.count = 0
+    hess.count = 0
+    hessp.count = 0
+    fun_constraint.count = 0
+    grad_constraint.count = 0
+    callback.count = 0
+
+    res = cyipopt.minimize_ipopt(fun, x0, method=method, args=(a0,),
+                                 kwargs={'b': b0}, **kwargs)
+
+    assert res.success
+    np.testing.assert_allclose(res.x[:2], [a0, b0], rtol=1e-3)
+
+    # confirm that the test covers what we think it does: all the functions
+    # that we provide are actually being executed; that is, the assertions
+    # are *passing*, not being skipped
+    assert fun.count > 0
+    if method in MINIMIZE_METHODS:
+        assert callback.count > 0
+    if method in jac_methods:
+        assert grad.count > 0
+    if method in hess_methods:
+        assert hess.count > 0
+    if method in constr_methods:
+        assert fun_constraint.count > 0
+        assert grad_constraint.count > 0
+        np.testing.assert_allclose(res.x[2:], [c0, d0], rtol=1e-3)
+    else:
+        np.testing.assert_allclose(res.x[2:], 0, atol=1e-3)
+
+    # For methods that support `hessp`, check that it works, too.
+    if method in hessp_methods:
+        del kwargs['hess']
+        kwargs['hessp'] = hessp
+
+        res = cyipopt.minimize_ipopt(fun, x0, method=method, args=(a0,),
+                                     kwargs={'b': b0}, **kwargs)
+        assert res.success
+        assert hessp.count > 0
+        np.testing.assert_allclose(res.x[:2], [a0, b0], rtol=1e-3)
+        if method in constr_methods:
+            np.testing.assert_allclose(res.x[2:], [c0, d0], rtol=1e-3)
+        else:
+            np.testing.assert_allclose(res.x[2:], 0, atol=1e-3)
```

Comment:
I'm curious why the tolerances in these checks are quite large.

Reply:
Think it was because of COBYLA. I'll see if they can be tighter for other methods.

Reply:
Yeah a few solvers have relative or absolute errors up to the order of `1e-5`. I tightened the tolerances to `5e-5` across the board and made them variables so they can be adjusted more easily. I don't think it's worth it to have different tolerances for each part of the test / every method, and `5e-5` is about the tightest we can go and have everything pass.

Reply:
Ok, that's fine.